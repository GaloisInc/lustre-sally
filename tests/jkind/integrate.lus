-- This file has 2 properties, both valid.

node integ(x : int) returns (sum : int);
let
  sum = x + (0 -> pre sum);
tel;

node main(x : int; y : int) returns (z : int);
var
  history : bool;
  prop1 : bool;
  prop2 : bool;
let
  z = integ(x);

  history = x > 0 and (true -> pre history);
  prop1 = history => z > 0;

  --%PROPERTY prop1;

  prop2 = integ(x) + integ(y) = integ(x + y);
  --%PROPERTY prop2;
tel;
