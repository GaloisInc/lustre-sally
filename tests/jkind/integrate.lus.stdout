Sally Model
===========

(define-state-type
   S
   (
     (history Bool)
     (integ Int)
     (prop1 Bool)
     (prop2 Bool)
     (x Int)
     (y Int)
     (z Int)
     (|gal-initializing| Bool)
     (|init_pre_history:init| Bool)
     (|init_pre_history| Bool)
     (|init_pre_integ:1:init| Bool)
     (|init_pre_integ:1| Int)
     (|init_pre_integ:2:init| Bool)
     (|init_pre_integ:2| Int)
     (|init_pre_integ:init| Bool)
     (|init_pre_integ| Int)
     (|init_pre_z:init| Bool)
     (|init_pre_z| Int)
     (|integ:1| Int)
     (|integ:2| Int)
     (|pre_history| Bool)
     (|pre_integ:1| Int)
     (|pre_integ:2| Int)
     (|pre_integ| Int)
     (|pre_z| Int)
     (|x:1| Int)
     (|x:2| Int)
     (|x:3| Int)
     (|x:4| Int))
   (
     (y Int)
     (|x:4| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |init_pre_z:init| true)
      (= |init_pre_history:init| true)
      (= |init_pre_integ:2:init| true)
      (= |init_pre_integ:1:init| true)
      (= |init_pre_integ:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|x:4| input.|x:4|)
      (= next.y input.y)
      (= next.|x:3| next.|x:4|)
      (= next.|pre_z| state.z)
      (ite
         state.|init_pre_z:init|
         (= next.|init_pre_z| 0)
         (= next.|init_pre_z| next.|pre_z|))
      (= next.|init_pre_z:init| false)
      (=
         next.z
         (+ next.|x:3| next.|init_pre_z|))
      (= next.|pre_history| state.history)
      (ite
         state.|init_pre_history:init|
         (= next.|init_pre_history| true)
         (= next.|init_pre_history| next.|pre_history|))
      (= next.|init_pre_history:init| false)
      (=
         next.history
         (and
            (< 0 next.|x:4|)
            next.|init_pre_history|))
      (=
         next.prop1
         (=>
            next.history
            (< 0 next.z)))
      (= next.|x:2| next.|x:4|)
      (= next.|pre_integ:2| state.|integ:2|)
      (ite
         state.|init_pre_integ:2:init|
         (= next.|init_pre_integ:2| 0)
         (= next.|init_pre_integ:2| next.|pre_integ:2|))
      (= next.|init_pre_integ:2:init| false)
      (=
         next.|integ:2|
         (+ next.|x:2| next.|init_pre_integ:2|))
      (= next.|x:1| next.y)
      (= next.|pre_integ:1| state.|integ:1|)
      (ite
         state.|init_pre_integ:1:init|
         (= next.|init_pre_integ:1| 0)
         (= next.|init_pre_integ:1| next.|pre_integ:1|))
      (= next.|init_pre_integ:1:init| false)
      (=
         next.|integ:1|
         (+ next.|x:1| next.|init_pre_integ:1|))
      (=
         next.x
         (+ next.|x:4| next.y))
      (= next.|pre_integ| state.integ)
      (ite
         state.|init_pre_integ:init|
         (= next.|init_pre_integ| 0)
         (= next.|init_pre_integ| next.|pre_integ|))
      (= next.|init_pre_integ:init| false)
      (=
         next.integ
         (+ next.x next.|init_pre_integ|))
      (=
         next.prop2
         (=
            (+ next.|integ:2| next.|integ:1|)
            next.integ))))

(query
   TS
   (or |gal-initializing| prop2))
(query
   TS
   (or |gal-initializing| prop1))

[Lustre] Validating properties:
[Lustre] Property prop2...
  [Sally][Valid] using 1 past state
[Lustre] Property prop1...
  [Sally][Valid] using 1 past state
[Lustre] Summary:
  [Valid] 2
  [Unknown] 0
  [Invalid] 0
[Lustre] Model status: [Valid]
